src/IECore/StringAlgo.cpp

```diff
@@ -95,6 +99,183 @@ bool matchInternal(
 	}
 }
 
+inline void substituteInternal( const char *s, const StringAlgo::VariableProvider &variables, std::string &result, const int recursionDepth, unsigned substitutions )
+{
+	if( recursionDepth > 8 )
+	{
+		throw IECore::Exception( "StringAlgo::substitute() : maximum recursion depth reached." );
+	}
+
+	while( *s )
+	{
+		switch( *s )
+		{
+			case '\\' :
+			{
+				if( substitutions & StringAlgo::EscapeSubstitutions )
+				{
+					s++;
+					if( *s )
+					{
+						result.push_back( *s++ );
+					}
+				}
+				else
+				{
+					// variable substitutions disabled
+					result.push_back( *s++ );
+				}
+				break;
+			}
+			case '$' :
+			{
+				if( substitutions & StringAlgo::VariableSubstitutions )
+				{
+					s++; // skip $
+					bool bracketed = *s =='{';
+					const char *variableNameStart = nullptr;
+					const char *variableNameEnd = nullptr;
+					if( bracketed )
+					{
+						s++; // skip initial bracket
+						variableNameStart = s;
+						while( *s && *s != '}' )
+						{
+							s++;
+						}
+						variableNameEnd = s;
+						if( *s )
+						{
+							s++; // skip final bracket
+						}
+					}
+					else
+					{
+						variableNameStart = s;
+						while( isalnum( *s ) )
+						{
+							s++;
+						}
+						variableNameEnd = s;
+					}
+
+					bool recurse = false;
+					const std::string &variable = variables.variable(
+						boost::string_view( variableNameStart, variableNameEnd - variableNameStart ),
+						recurse
+					);
+
+					if( recurse )
+					{
+						substituteInternal( variable.c_str(), variables, result, recursionDepth + 1, substitutions );
+					}
+					else
+					{
+						result += variable;
+					}
+				}
+				else
+				{
+					// variable substitutions disabled
+					result.push_back( *s++ );
+				}
+				break;
+			}
+			case '#' :
+			{
+				if( substitutions & StringAlgo::FrameSubstitutions )
+				{
+					int padding = 0;
+					while( *s == '#' )
+					{
+						padding++;
+						s++;
+					}
+					std::ostringstream padder;
+					padder << std::setw( padding ) << std::setfill( '0' ) << variables.frame();
+					result += padder.str();
+				}
+				else
+				{
+					// frame substitutions disabled
+					result.push_back( *s++ );
+				}
+				break;
+			}
+			case '~' :
+			{
+				if( substitutions & StringAlgo::TildeSubstitutions && result.size() == 0 )
+				{
+					if( const char *v = getenv( "HOME" ) )
+					{
+						result += v;
+					}
+					++s;
+					break;
+				}
+				else
+				{
+					// tilde substitutions disabled
+					result.push_back( *s++ );
+				}
+				break;
+			}
+			default :
+				result.push_back( *s++ );
+				break;
+		}
+	}
+}
+
+struct CompoundDataVariableProvider : public StringAlgo::VariableProvider
+{
+
+	CompoundDataVariableProvider( const CompoundData *variables )
+		:	m_variables( variables )
+	{
+	}
+
+	int frame() const override
+	{
+		const IECore::IntData *d = m_variables->member<IECore::IntData>( "frame" );
```

Comment:
In the Context implementation we handle float frames by rounding. Seems like that could be desirable here too.

Reply:
8592d4852e6bb53f86a04e7e77d1cfe1218115f1 (supported by aedd0eb510d5025404147c80768fe05329f64b10).